ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/fullsync/FullSynchronizer.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.ethereum.eth.sync.fullsync;
+
+import tech.pegasys.pantheon.ethereum.ProtocolContext;
+import tech.pegasys.pantheon.ethereum.eth.manager.EthContext;
+import tech.pegasys.pantheon.ethereum.eth.sync.BlockBroadcaster;
+import tech.pegasys.pantheon.ethereum.eth.sync.ChainDownloader;
+import tech.pegasys.pantheon.ethereum.eth.sync.SynchronizerConfiguration;
+import tech.pegasys.pantheon.ethereum.eth.sync.TrailingPeerRequirements;
+import tech.pegasys.pantheon.ethereum.eth.sync.state.PendingBlocks;
+import tech.pegasys.pantheon.ethereum.eth.sync.state.SyncState;
+import tech.pegasys.pantheon.ethereum.mainnet.ProtocolSchedule;
+import tech.pegasys.pantheon.metrics.MetricsSystem;
+
+public class FullSynchronizer<C> {
+
+  private final BlockPropagationManager<C> blockPropagationManager;
+  private final ChainDownloader chainDownloader;
+  private final SynchronizerConfiguration syncConfig;
+  private final ProtocolContext<C> protocolContext;
+  private final SyncState syncState;
+
+  public FullSynchronizer(
+      final SynchronizerConfiguration syncConfig,
+      final ProtocolSchedule<C> protocolSchedule,
+      final ProtocolContext<C> protocolContext,
+      final EthContext ethContext,
+      final SyncState syncState,
+      final MetricsSystem metricsSystem) {
+    this.syncConfig = syncConfig;
+    this.protocolContext = protocolContext;
+    this.syncState = syncState;
+
+    this.blockPropagationManager =
```

Comment:
I think the terminology here is getting a bit fuzzy.  I think the "FullSynchronizer" is really a "FullDownloaderFactory" (similarly "FastSynchronizer" -> "FastDownloaderFactory").  And I think that the `blockPropagationManager` should belong to the `Synchronizer`.  In my mind, the synchronizer encapsulates all things related to staying up-to-date with the network:  tracking peer stats, managing propagated blocks, downloading from the network.  And full / fast modes are only concerned with how we download data from the network.

Reply:
Yep, I can see that working.  Done.